Add fake shaders, affine/linear transformations and convolution matrix to GPU shaders

TheEt1234 opened the pull request:
All of the features are documented in docs/gpu.txt

## Fake shaders
- Performance: Shaders take about 5~20ms for me (when i run the RGB showcase)
- Security
  - Shaders are limited in almost the same way as mesecons luacontroller sandboxes (debug hooks to control the amount of time that shaders can take up, setfenv to control their environment, strings aren't sandboxed because the shader sandbox can't use strings)
  - They should be more secure than mesecons luacontrollers as they don't allow access to strings, so they don't have the weakness of string concatination causing lag

## Convolution matrix
- Performance: Takes about 10-25ms for a 64x64 image, with a 5x5 matrix for me
  - It could potentially be even faster if digistuff GPUs just stored colors in plain numbers, should I contribute that? If I did I would also make them store the buffer in just a plain array (like ``buffer[y * xsize + x] = color`` I probably messed that up but you get the idea), but everything could potentially be faster if that was the case

## Affine/linear transformations
- Performance: 5~10ms if not interpolated, 10-20ms when interpolated but tends to stabilize at 5~10ms?
   - Also when testing this I discovered that most things are MUCH faster (like 0~2ms) when my luanti window is unfocused, weird, I am not sure if this is a byproduct of me measuring things with os.clock, or if it's geniuenly just faster and that I shouldn't care 

### Notes
- All the performance testing was done with lightly modified versions of the showcase code, and with my laptop's power profile being power saving, so even a cheap server will probably be faster than that
- The invert_3x3_matrix and invert_2x2 matrix functions were done by chatGPT (I was having gaps in the buffer when scaling so I asked chatgpt about them and it suggested inverting the matrix and doing it a slightly different way)

### Showcase:
<img width="656" height="441" alt="image" src="https://github.com/user-attachments/assets/dfb7c8d3-309f-4772-b29c-209e006eef02" />
<details>
<summary>Showcase code:</summary>

```lua
--interrupt(.5)
local shader_cmd = {command="fakeshader", buffer = 1,
code = [[
local r = 16 -- radius
local dist = 10 -- distance between circles
local cX, cY = x-(xsize/2), y-(ysize/2) -- center X, center Y
local r2 = r^2

local red_circle   = r2 - (cX^2+(cY+dist)^2) -- thanks pythagoras
local green_circle = r2 - ((cX+dist)^2+cY^2)
local blue_circle  = r2 - ((cX-dist)^2+cY^2)

-- make them solid
red_circle, green_circle, blue_circle =
 math.min(1, red_circle)*255,
 math.min(1, green_circle)*255,
 math.min(1, blue_circle)*255

return red_circle, green_circle, blue_circle
]]}


local a,b,c = -1, .3/8, 4
local convolution_matrix = {
{0,0,a,0,0},
{0,b,b,b,0},
{a,b,c,b,a},
{0,b,b,b,0},
{0,0,a,0,0},
}

local angle = math.pi/4

local transform_by = {
{math.cos(angle), -math.sin(angle),0},
{math.sin(angle), math.cos(angle),0},
{-.01,0,0.8}
}

digiline_send("gpu",{
   {command ="createbuffer",buffer=1, xsize = 64, ysize = 64, fill = "#000000"},
   shader_cmd,
{
command = "convolutionmatrix",
buffer = 1,
matrix = convolution_matrix,
},
{
command = "transform",
buffer = -1,
matrix = transform_by,
x1=1,
y1=1,
x2=64,
y2=64,
center_x = 32,
center_y = 32,
interpolation=false
},

   {command = "sendregion", buffer=1, channel = "1", x1=1,y1=1,x2=64,y2=64},

})
```
</details>

<br>
Also, would it be ok to remove permanent saving of digistuff GPU buffers to node meta and replace that with just a temporary table outside of the GPU? I am not sure if using minetest.serialize on GPU buffers is a good idea because it has to build up the `return { { "ffffff", "aaaaa", ...}, ... }` which I *assume* isn't performant...

TheEt1234 committed:
Add fake shaders to digistuff GPU

TheEt1234 committed:
Add convolution matrix to digistuff GPUs

TheEt1234 committed:
Add transform to digistuff GPU

TheEt1234 committed:
Now it should be ready

OgelGames commented:
```
Checking gpu.lua                                  10 warnings

    gpu.lua:481:27: accessing undefined field round of global math
    gpu.lua:481:42: accessing undefined field round of global math
    gpu.lua:825:13: variable errmsg was previously defined on line 766
    gpu.lua:838:9: shadowing definition of variable xsize on line 501
    gpu.lua:838:16: shadowing definition of variable ysize on line 501
    gpu.lua:871:9: shadowing definition of variable x1 on line 501
    gpu.lua:871:13: shadowing definition of variable y1 on line 501
    gpu.lua:871:17: shadowing definition of variable x2 on line 501
    gpu.lua:871:21: shadowing definition of variable y2 on line 501
    gpu.lua:903:12: shadowing definition of variable color on line 502
```

TheEt1234 committed:
Fix complaints from mr luacheck

TheEt1234 commented:
> ```
> Checking gpu.lua                                  10 warnings
> 
>     gpu.lua:481:27: accessing undefined field round of global math
>     gpu.lua:481:42: accessing undefined field round of global math
>     gpu.lua:825:13: variable errmsg was previously defined on line 766
>     gpu.lua:838:9: shadowing definition of variable xsize on line 501
>     gpu.lua:838:16: shadowing definition of variable ysize on line 501
>     gpu.lua:871:9: shadowing definition of variable x1 on line 501
>     gpu.lua:871:13: shadowing definition of variable y1 on line 501
>     gpu.lua:871:17: shadowing definition of variable x2 on line 501
>     gpu.lua:871:21: shadowing definition of variable y2 on line 501
>     gpu.lua:903:12: shadowing definition of variable color on line 502
> ```

Fixed

cheapie commented:
Ooh, affine transformations are something I've been wanting to implement for a long time now but I've never had any clue where to start on them. The rest I admittedly don't care about so much, but I tend to be of the KDE philosophy here that more features are usually a good thing (as long as they work and don't get in the way).

> Also, would it be ok to remove permanent saving of digistuff GPU buffers to node meta and replace that with just a temporary table outside of the GPU?

I understand the concern here, but at the same time I have a lot of builds that depend on buffers being persistent across server restarts. I wouldn't be too opposed to saving them to node meta requiring an explicit command to do so though, I can go through those builds and add that where needed.

I'll try to test this PR sometime when I have time, maybe this weekend, not sure at the moment.

TheEt1234 commented:
> > Also, would it be ok to remove permanent saving of digistuff GPU buffers to node meta and replace that with just a temporary table outside of the GPU?
> 
> I understand the concern here, but at the same time I have a lot of builds that depend on buffers being persistent across server restarts. I wouldn't be too opposed to saving them to node meta requiring an explicit command to do so though, I can go through those builds and add that where needed.

What if buffers were saved in a table outside of the GPU, but that table was also saved to a file (like how pipeworks lua entities get saved), and the GPU would first try to read from node meta to see if there is an old buffer stored there

Also would it be OK to make all the colors in the GPU be numbers instead of strings in this PR or would that be too much to test for (should I do that in another PR or should I not do that, the reason I want to do that is I think it would be slightly faster)

cheapie commented:
> What if buffers were saved in a table outside of the GPU, but that table was also saved to a file (like how pipeworks lua entities get saved), and the GPU would first try to read from node meta to see if there is an old buffer stored there

I guess from the point of view of the user (which is where my concerns came from here) it doesn't really matter where it's saved as long as it is saved somewhere. Probably shouldn't be a file though, maybe mod storage if you think that'll be faster?

cheapie commented:
OK, so I tested this some - it seems to mostly work, but I did run into a crash:

```
AsyncErr: Lua: Runtime error from mod 'mesecons' in callback environment_Step(): ....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:186: bad argument #1 to 'tonumber' (string expected, got nil)
stack traceback:
	[C]: in function 'tonumber'
	....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:186: in function 'unpack_color'
	....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:207: in function 'blend'
	....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:679: in function 'runcommand'
	....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:986: in function 'action'
[snip digilines and stuff getting here in the first place]
```

This is the code I was running in the Luacontroller for testing, it's a bit of a mess since it grew over time as I added more effects on top of each other, but it was working until line 18 was added:

```lua
if event.type == "program" then
	mem.angle = 0
	mem.step = 0
	digiline_send("gpu",{
		{command="createbuffer",buffer=0,xsize=64,ysize=64},
		{command="createbuffer",buffer=1,xsize=64,ysize=64},
		{command="fakeshader",buffer=0,code="return (x/xsize)*255,math.random(0,255),0"},
	})
	interrupt(0,"tick")
elseif event.iid == "tick" then
	local angle = mem.angle/180*math.pi
	local scale = math.sin(mem.step)*0.3+0.6
	digiline_send("disp",tostring(mem.angle))
	digiline_send("gpu",{
		{command="copy",src=0,dst=1,srcx=1,srcy=1,dstx=1,dsty=1,xsize=64,ysize=64},
		{command="transform",buffer=1,x1=1,y1=1,x2=64,y2=64,center_x=32,center_y=32,interpolation=false,matrix={{scale,0},{0,scale}}},
		{command="transform",buffer=1,x1=1,y1=1,x2=64,y2=64,center_x=32,center_y=32,interpolation=false,matrix={{math.cos(angle),-math.sin(angle)},{math.sin(angle),math.cos(angle)}}},
		{command="fakeshader",buffer=1,code="return r,g,255*math.sin(x/2)"}, --This is line 18, the crash happened after this was added
		{command="send",buffer=1,channel="screen"},
	})
	mem.angle = (mem.angle%360)+10
	mem.step = mem.step+0.1
	interrupt(0.1,"tick")
end
```

OgelGames commented:
With all these big features being added, do you think the GPU should be split into its own mod? More than a third of the code in digistuff is for the GPU and nothing in digistuff depends on it.

SwissalpS commented:
> With all these big features being added, do you think the GPU should be split into its own mod? More than a third of the code in digistuff is for the GPU and nothing in digistuff depends on it.

I second that idea.
Only downside I see is that then there would have to be aliases or LBMs to convert existing to new mod. Aliases aren't super with recipes :/

TheEt1234 committed:
Fix bug with shaders

TheEt1234 commented:
> OK, so I tested this some - it seems to mostly work, but I did run into a crash:
> 
> ```
> AsyncErr: Lua: Runtime error from mod 'mesecons' in callback environment_Step(): ....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:186: bad argument #1 to 'tonumber' (string expected, got nil)
> stack traceback:
> 	[C]: in function 'tonumber'
> 	....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:186: in function 'unpack_color'
> 	....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:207: in function 'blend'
> 	....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:679: in function 'runcommand'
> 	....minetest/worlds/techagetest/worldmods/digistuff/gpu.lua:986: in function 'action'
> [snip digilines and stuff getting here in the first place]
> ```
> 
> This is the code I was running in the Luacontroller for testing, it's a bit of a mess since it grew over time as I added more effects on top of each other, but it was working until line 18 was added:
> 
> ```lua
> if event.type == "program" then
> 	mem.angle = 0
> 	mem.step = 0
> 	digiline_send("gpu",{
> 		{command="createbuffer",buffer=0,xsize=64,ysize=64},
> 		{command="createbuffer",buffer=1,xsize=64,ysize=64},
> 		{command="fakeshader",buffer=0,code="return (x/xsize)*255,math.random(0,255),0"},
> 	})
> 	interrupt(0,"tick")
> elseif event.iid == "tick" then
> 	local angle = mem.angle/180*math.pi
> 	local scale = math.sin(mem.step)*0.3+0.6
> 	digiline_send("disp",tostring(mem.angle))
> 	digiline_send("gpu",{
> 		{command="copy",src=0,dst=1,srcx=1,srcy=1,dstx=1,dsty=1,xsize=64,ysize=64},
> 		{command="transform",buffer=1,x1=1,y1=1,x2=64,y2=64,center_x=32,center_y=32,interpolation=false,matrix={{scale,0},{0,scale}}},
> 		{command="transform",buffer=1,x1=1,y1=1,x2=64,y2=64,center_x=32,center_y=32,interpolation=false,matrix={{math.cos(angle),-math.sin(angle)},{math.sin(angle),math.cos(angle)}}},
> 		{command="fakeshader",buffer=1,code="return r,g,255*math.sin(x/2)"}, --This is line 18, the crash happened after this was added
> 		{command="send",buffer=1,channel="screen"},
> 	})
> 	mem.angle = (mem.angle%360)+10
> 	mem.step = mem.step+0.1
> 	interrupt(0.1,"tick")
> end
> ```

Ok, so turns out i didn't do enough testing with the shaders, whoops! should be fixed now

> `{command="fakeshader",buffer=1,code="return r,g,255*math.sin(x/2)"}`

The code for this shader is wrong, it should've been `return color.r, color.g, 255*math.sin(x/2)` instead (should i change the environment of the shaders to use `r, g, b` instead of `color.r, color.g, color.b` for the color)

What should've happened: The GPU should've sent a message (on the channel "gpu") telling you that you returned wrong the wrong values (`nil, nil, <whatever the value of 255*math.sin(x/2) was>`)

What actually happened: 
in line 329 of gpu.lua, the run_shader function returned false with the error message, but it should've *`error()`*'d instead

around line 360 in gpu.lua (the `local ok, errmsg = pcall(` line), `errmsg` was false and a third ignored return was the actual error message

and since nothing was done to the new buffer, the new buffer was `{{}, {}, {} ,... }`, and that's how you got this error

TheEt1234 commented:
> With all these big features being added, do you think the GPU should be split into its own mod? More than a third of the code in digistuff is for the GPU and nothing in digistuff depends on it.

I think this would be ok if:
1) All player made creations kept working (Maybe digistuff would have to be a modpack?)
2) There was a good reason for doing that

> big features

This PR looks big but that's probably just because of my formatter

cheapie commented:
It seems to work properly now after that most recent commit, at least for me and with a fixed version of that same test program. I think my only current problems with it, unless I find more, are documentation things:

- gpu.txt line 201 suggests that the current pixel color is available in r/g/b instead of color.r/color.g/color.b. I don't really have a preference as to which one it is so long as the documentation matches it.
- The section on transforms (gpu.txt lines 276-279) states that the position range is 0-64 when it's actually 1-64

As far as the other things discussed here so far go, my thoughts on them:

- Changing it to use some other storage method for the buffers internally - I don't really see much of a problem with changing it to something like mod storage if it's indeed faster, probably best done in a separate PR though.
- Changing the color representation to numbers - similar story here, if nothing user-visible changes (still accepts strings in commands and converts them, send/sendregion still output tables of tables of strings) then I see no downside to changing an internal representation to something faster. Probably better done as a separate PR (same one as the previous item?)
- Splitting the GPU into its own mod - I'm mildly against this, but only mildly. It wouldn't be hard to split out (one alias should do it for compatibility with existing worlds) but I'm not convinced there's any real reason to do so.

TheEt1234 committed:
Fix slightly wrong position range in digistuff gpu documentation

TheEt1234 committed:
Slightly improve documentation for shaders

TheEt1234 commented:
> gpu.txt line 201 suggests that the current pixel color is available in r/g/b instead of color.r/color.g/color.b. I don't really have a preference as to which one it is so long as the documentation matches it.

`r, g, b = get_color(x, y)` is supposed to be how you use the function

I changed the way to get the current color from `color.r, color.g, color.b` to just `r, g, b`, because it should be *somewhat slightly* faster and it feels more consistent, the new documentation for that is

```
- r: The red value of the current pixel being processed
- g: The green value of the current pixel being processed
- b: The blue value of the current pixel being processed
```


I also changed the documentation for `get_color`:
```
- get_color(x, y): Allows you to get a color from the buffer, returns 3 values: r, g, b
```

> The section on transforms (gpu.txt lines 276-279) states that the position range is 0-64 when it's actually 1-64

Thanks, I fixed it

TheEt1234 committed:
Change shader API slightly (accessing color went from color.r/color.g/color.b to just r/g/b)

TheEt1234 committed:
Remove useless unused local

TheEt1234 committed:
Disallow bytecode in fakeshaders (in case someone turned off mod security, don't do that)

SwissalpS commented on gpu.lua:
```diff
@@ -758,6 +758,8 @@ local function runcommand(pos, meta, command)
 		if type(command.code) ~= "string" then return false, "No code provided" end
 		if #command.code > MAX_SHADER_CODE_LENGTH then return false, "Code too large" end
 
+		if command.code:byte(1) == 27 then return false, "No bytecode!" end -- This is technically not needed, but if someone turned off mod security (don't do that!!!!) then someone could abuse bytecode
```
Would it be possible to put the comment above the line it corresponds to?
Also one exclamation mark is sufficient.

TheEt1234 replied:
Fixed!

also "Allow edits by maintainers" is checked, so you can push trivial changes like that directly to my fork I think

TheEt1234 committed:
Fix comment

wsor4035 reviewed (approved):
I'm passing through cheapies approval, and a minimal review of my own
